Updates for NavMenu, startplugin, and docs (#4419)
nautobot/nautobot: nautobot/docs/development/core/index.md

For more information, please visit [SemVer.org](https://semver.org/).

> #### Public Interface

Author: This is under `Release Management -> Versioning -> Public Interface` to define what is SemVer'd vs not. Taking inspiration from here: https://regret.readthedocs.io/en/stable/before-you-deprecate/#document-your-public-api, I think this is the correct place for this. 

This is simply saying, "hey, not every docstring and private function is semver'd, we reserve the right to change the function signature on a private function during every patch release, which to me belongs under release management.

The guiding principals behind Nautobot's SemVer strategy is scoped to adhering to the prescribed public interface. This includes both the application operators and application developer. Anything not explicitly covered here is presumed to not be covered under SemVer.

> * The REST API interface within a Version, will not change, but may add to

Reviewer: ```suggestion
* Behavior of a given X.Y version of the REST API. New versions of the REST API may be added, which may introduce changes to
```

    * URL structures
    * Expected request and response formats
    * Behavior of each endpoint
    * Status and error codes
    * The HTTP verbs

> * Graphql data structures

Reviewer: ```suggestion
* GraphQL data structures
```

* URL structures for primary navigation and feature access

> * Any Django model and their relationships for models provided by Nautobot

Reviewer: I don't have specific verbiage to suggest as an alternative, but this feels a bit vague to me. We absolutely might introduce new fields or new behaviors to Django models.

Author: i added an info statement below, does that help?

* Functions, classes, and modules in `nautobot.apps.*` and their associated signature
* Any management commands provided by Nautobot
* Documented configuration options

> While anything not explicitly stated above should be not be considered scoped to SemVer, the following should be strictly avoided.

Reviewer: ```suggestion
While anything not explicitly stated above should be not be considered scoped to SemVer, the following should be strictly avoided when developing non-core Nautobot Apps and Jobs.
```

* Functions or methods that start with an underscore (e.g., `_private_func`)
* Any functions, classes, and modules **not** in `nautobot.apps.*`.
* Any modules, classes, functions, or feature that are explicitly marked as "internal" or "experimental" in the documentation

> * Any features or functionalities that aren't part of the official documentation should not be relied upon.

Reviewer: This bullet list, at least, still feels like it belongs in the `apps` and `jobs` development documentation rather than here. Maybe at a minimum those docs should cross-link to this section?
